MacOS compatibility (again)

Roboe opened the pull request:
(Apparently GitHub doesn't know how to reopen a Pull Request, so I'm recreating #7. Meh.)

**--> Must fix #4 once before merging. <--**

This PR specifically replaces `mktemp` with `gmktemp`. `gmktemp` is available in GNU coreutils (`$ brew install coreutils`).
I could have written the command like `mktemp -u` instead of `mktemp --dry-run`, but I decided to keep readability because we already need `coreutils` to provide `gsha256sum`.  

Please, @jfdhuiz, could you test this on your machine?

Roboe committed:
MacOS: mktemp -> gmktemp

Thanks @jfdhuiz

Roboe committed:
Add MacOS requirements to README.md

z3ntu commented on Makefile:
```diff
@@ -2,7 +2,19 @@
+CURL        := $(shell command -v curl 2>&1)
+ZIP         := $(shell command -v zip 2>&1)
+UNZIP       := $(shell command -v unzip 2>&1)
+ifeq ($(uname -s),"Darwin")
```
This comparison doesn't seem to work at all as on macOS the `else` path is taken.

jfdhuiz replied:
$ if [ $(uname -s) = 'Darwin' ] ; then echo true; else echo false ; fi
true
$ if [ $(uname -s) = 'Linux' ] ; then echo true; else echo false ; fi
false

(On a Mac)

z3ntu committed:
Fix comparison, sort zip order, use GNU cp

z3ntu commented:
As of the latest commit I pushed, the resulting zip has the same checksum built on both Linux and macOS.
![image](https://user-images.githubusercontent.com/3768500/48855138-c67fe700-edb3-11e8-84bf-37aff596f55d.png)


`brew install coreutils findutils` is needed I think

z3ntu committed:
Adjust README again

Roboe commented:
I rebased on top of master and fixed a git conflict caused by a linebreak.

Your work LGTM, @z3ntu. Could you try it another time in your machines? Feel free to merge, tag and release the new flashable ZIP then with #14 changes, I'lI probably be absent from GitHub for a while once again.

```
$ make clean && make release
[...]
$ cat release/modem-19.02.1_2019-03-03.zip.sha256sum 
7892548bdb447c003f6a545f9cc19d7019cbf64f05c1263cc6f94d9fe075a257  modem-19.02.1_2019-03-03.zip
```

(I did multiple executions in Linux, same hash)

z3ntu commented:
```
cat release/modem-19.02.1_2019-03-03.zip.sha256sum
7da5c204a2103245661dd06e331aeaec314ba5f3ebb5b18b28bb6639ecce6c64  modem-19.02.1_2019-03-03.zip
```
Could you upload your zip so I can look for differences between yours and mine?

Roboe commented:
Of course. There you go:

https://send.firefox.com/download/c48c81eb9d/#5q4RtZPPsw6jUvzbPVl7oA
(It's bigger than 10MB, so GitHub refuses to host it) 

```
7892548bdb447c003f6a545f9cc19d7019cbf64f05c1263cc6f94d9fe075a257  modem-19.02.1_2019-03-03.zip
```
(built at 00c4f4a51ee76cf9c26a85c7b0aa73d88bead9b6 on an up-to-date Debian Testing "buster" machine)

z3ntu commented:
What distro are you using or what's the output of `zip -v | head` ?

Roboe commented:
> on an up-to-date Debian Testing "buster" machine

```
$ zip -v | head
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.
Currently maintained by E. Gordon.  Please send bug reports to
the authors using the web page at www.info-zip.org; see README for details.

Latest sources and executables are at ftp://ftp.info-zip.org/pub/infozip,
as of above date; see http://www.info-zip.org/ for other sites.

Compiled with gcc 6.3.0 20170221 for Unix (Linux ELF).
```

z3ntu commented:
Built from **this branch** on Linux I get 
```
5ec0684f73637e2a8b3ba578f595b7417bb8d389a0a4197a70e25527152ba040  modem-19.02.1_2019-03-03.zip
```
Could you try that maybe?

Roboe commented:
> (built at 00c4f4a51ee76cf9c26a85c7b0aa73d88bead9b6 on an up-to-date Debian Testing "buster" machine)

This branch latest commit is 00c4f4a51ee76cf9c26a85c7b0aa73d88bead9b6. Did you `make clean && make release`? Maybe GNUMake is missing the `sdi.mbn` binary when unpacking the firmware.

z3ntu commented:
It seems that we lost the functionality that an error is thrown when a dependency isn't found with #6 ...

z3ntu commented:
I also get `5ec0684f73637e2a8b3ba578f595b7417bb8d389a0a4197a70e25527152ba040` on macOS using this branch

Roboe commented:
> It seems that we lost the functionality that an error is thrown when a dependency isn't found with #6 ...

Which dependency are your referring to? If a target can't be build/download/unpacked, Make halts and you get the shell error, no?

> I also get `5ec0684f73637e2a8b3ba578f595b7417bb8d389a0a4197a70e25527152ba040` on macOS using this branch

🤯 :man_shrugging:

z3ntu commented:
With dependency I mean something like `curl` or `sha256sum` and no, manual error handling would be needed there (so a check if each variable is defined and if not `$(error message)`)
e.g. it looks like
```
root@b770552f0e7c:/modem_zip_generator# make release
Makefile:88: warning: overriding recipe for target 'updates/19.02.1-sibon-dc48370a-ota.zip'
Makefile:82: warning: ignoring old recipe for target 'updates/19.02.1-sibon-dc48370a-ota.zip'
Downloading 19.02.1-sibon-dc48370a-ota.zip...
mkdir: created directory 'updates'
/bin/bash: progress-bar: command not found
make: [Makefile:90: updates/19.02.1-sibon-dc48370a-ota.zip] Error 127 (ignored)
/usr/bin/sha256sum: updates/19.02.1-sibon-dc48370a-ota.zip: No such file or directory
updates/19.02.1-sibon-dc48370a-ota.zip: FAILED open or read
/usr/bin/sha256sum: WARNING: 1 listed file could not be read
Unpacking firmware files...
/bin/bash: line 1: j: command not found
make: [Makefile:105: src/firmware-update/] Error 127 (ignored)
Unpacking Edify interpreter...
/bin/bash: line 1: j: command not found
make: [Makefile:96: src/META-INF/com/google/android/update-binary] Error 127 (ignored)
/usr/bin/cp: cannot stat './src/firmware-update/': No such file or directory
make: *** [Makefile:64: build/modem.zip] Error 1
```

z3ntu commented:
```
$ docker run --rm -it debian:testing /bin/bash
apt-get update && apt-get install git make curl zip unzip
git clone https://github.com/WeAreFairphone/modem_zip_generator -b feature/macos-compatibility
cd modem_zip_generator
make release
cat release/modem-19.02.1_2019-03-03.zip.sha256sum 
```
results in
```
5ec0684f73637e2a8b3ba578f595b7417bb8d389a0a4197a70e25527152ba040  modem-19.02.1_2019-03-03.zip
```

Roboe commented:
I'm recreating the build in another machine I have my hands on. 

Edit: Whoops.

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic

$ zip -v | head
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.
Currently maintained by E. Gordon.  Please send bug reports to
the authors using the web page at www.info-zip.org; see README for details.

Latest sources and executables are at ftp://ftp.info-zip.org/pub/infozip,
as of above date; see http://www.info-zip.org/ for other sites.

Compiled with gcc 6.3.0 20170415 for Unix (Linux ELF).

$ cat release/modem-19.02.1_2019-03-03.zip.sha256sum 
c62c5a0b7f409e3fe3d3935cecbbe019c1a183f1c39f53aaefb74d9e052a40a3  modem-19.02.1_2019-03-03.zip
```

Roboe commented:
Let's see what diffoscope says comparing my different ZIPs.

Edit: Seems to be an issue with file permissions, 644 on Debian vs 664 on Ubuntu, if I got it correctly.

```
$ diffoscope release/modem-19.02.1_2019-03-03.debian.zip release/modem-19.02.1_2019-03-03.ubuntu.zip
 |############################################################################################################################################|  100%                             Time: 0:00:05 
--- release/modem-19.02.1_2019-03-03.debian.zip
+++ release/modem-19.02.1_2019-03-03.ubuntu.zip
├── zipinfo /dev/stdin
│ @@ -1,11 +1,11 @@
│  Zip file size: 33300827 bytes, number of entries: 9
│ --rw-r--r--  3.0 unx   549068 b- defN 80-Jan-01 00:00 firmware-update/emmc_appsboot.mbn
│ --rw-r--r--  3.0 unx 58967552 b- defN 80-Jan-01 00:00 firmware-update/NON-HLOS.bin
│ --rw-r--r--  3.0 unx   191452 b- defN 80-Jan-01 00:00 firmware-update/rpm.mbn
│ --rw-r--r--  3.0 unx   280944 b- defN 80-Jan-01 00:00 firmware-update/sbl1.mbn
│ --rw-r--r--  3.0 unx    11712 b- defN 80-Jan-01 00:00 firmware-update/sdi.mbn
│ --rw-r--r--  3.0 unx  6221312 b- defN 80-Jan-01 00:00 firmware-update/splash.img
│ --rw-r--r--  3.0 unx   342344 b- defN 80-Jan-01 00:00 firmware-update/tz.mbn
│ --rw-r--r--  3.0 unx   865492 b- defN 80-Jan-01 00:00 META-INF/com/google/android/update-binary
│ --rw-r--r--  3.0 unx     1253 t- defN 80-Jan-01 00:00 META-INF/com/google/android/updater-script
│ +-rw-rw-r--  3.0 unx   549068 b- defN 80-Jan-01 00:00 firmware-update/emmc_appsboot.mbn
│ +-rw-rw-r--  3.0 unx 58967552 b- defN 80-Jan-01 00:00 firmware-update/NON-HLOS.bin
│ +-rw-rw-r--  3.0 unx   191452 b- defN 80-Jan-01 00:00 firmware-update/rpm.mbn
│ +-rw-rw-r--  3.0 unx   280944 b- defN 80-Jan-01 00:00 firmware-update/sbl1.mbn
│ +-rw-rw-r--  3.0 unx    11712 b- defN 80-Jan-01 00:00 firmware-update/sdi.mbn
│ +-rw-rw-r--  3.0 unx  6221312 b- defN 80-Jan-01 00:00 firmware-update/splash.img
│ +-rw-rw-r--  3.0 unx   342344 b- defN 80-Jan-01 00:00 firmware-update/tz.mbn
│ +-rw-rw-r--  3.0 unx   865492 b- defN 80-Jan-01 00:00 META-INF/com/google/android/update-binary
│ +-rw-rw-r--  3.0 unx     1253 t- defN 80-Jan-01 00:00 META-INF/com/google/android/updater-script
│  9 files, 67431129 bytes uncompressed, 33299597 bytes compressed:  50.6%
```

Roboe commented:
`umask` makes my machines to come to an agreement. I also flushed the `updates/` folder on both machines to force them to re-download Fairphone's ZIPs (futile, since the hash matched, but anyway).

```
$ umask 077 && make clean && make release
[...]
$ cat release/modem-19.02.1_2019-03-03.zip.sha256sum 
99bcafc9aac1529a7271f3ce1021d8f893d90db7865cdb5fafc6f89e114f00b1  modem-19.02.1_2019-03-03.zip
```

BUT the thing is our machines still don't get along.

z3ntu commented:
Here's my zip so you can compare yourself: https://send.firefox.com/download/95269879ee/#szUO7FX03_usEY838m07Wg

awesome, that you can tell firefox send to allow more than 1 download now

Roboe commented:
Interactive: https://try.diffoscope.org/ethdbuyghkge.html

```
--- a/modem-19.02.1_2019-03-03.zip
+++ b/modem-19.02.1_2019-03-03.z3ntu.zip
├── zipinfo /dev/stdin
│ @@ -1,11 +1,11 @@
│  Zip file size: 33300827 bytes, number of entries: 9
│ --rw-------  3.0 unx   549068 b- defN 80-Jan-01 00:00 firmware-update/emmc_appsboot.mbn
│ --rw-------  3.0 unx 58967552 b- defN 80-Jan-01 00:00 firmware-update/NON-HLOS.bin
│ --rw-------  3.0 unx   191452 b- defN 80-Jan-01 00:00 firmware-update/rpm.mbn
│ --rw-------  3.0 unx   280944 b- defN 80-Jan-01 00:00 firmware-update/sbl1.mbn
│ --rw-------  3.0 unx    11712 b- defN 80-Jan-01 00:00 firmware-update/sdi.mbn
│ --rw-------  3.0 unx  6221312 b- defN 80-Jan-01 00:00 firmware-update/splash.img
│ --rw-------  3.0 unx   342344 b- defN 80-Jan-01 00:00 firmware-update/tz.mbn
│ --rw-------  3.0 unx   865492 b- defN 80-Jan-01 00:00 META-INF/com/google/android/update-binary
│ --rw-------  3.0 unx     1253 t- defN 80-Jan-01 00:00 META-INF/com/google/android/updater-script
│ +-rw-r--r--  3.0 unx   865492 b- defN 80-Jan-01 00:00 META-INF/com/google/android/update-binary
│ +-rw-r--r--  3.0 unx     1253 t- defN 80-Jan-01 00:00 META-INF/com/google/android/updater-script
│ +-rw-r--r--  3.0 unx 58967552 b- defN 80-Jan-01 00:00 firmware-update/NON-HLOS.bin
│ +-rw-r--r--  3.0 unx   549068 b- defN 80-Jan-01 00:00 firmware-update/emmc_appsboot.mbn
│ +-rw-r--r--  3.0 unx   191452 b- defN 80-Jan-01 00:00 firmware-update/rpm.mbn
│ +-rw-r--r--  3.0 unx   280944 b- defN 80-Jan-01 00:00 firmware-update/sbl1.mbn
│ +-rw-r--r--  3.0 unx    11712 b- defN 80-Jan-01 00:00 firmware-update/sdi.mbn
│ +-rw-r--r--  3.0 unx  6221312 b- defN 80-Jan-01 00:00 firmware-update/splash.img
│ +-rw-r--r--  3.0 unx   342344 b- defN 80-Jan-01 00:00 firmware-update/tz.mbn
│  9 files, 67431129 bytes uncompressed, 33299597 bytes compressed:  50.6%
├── zipnote {}
│ @@ -1,13 +1,19 @@
│ -Filename: firmware-update/emmc_appsboot.mbn
│ +Filename: META-INF/com/google/android/update-binary
│ +Comment: 
│ +
│ +Filename: META-INF/com/google/android/updater-script
│  Comment: 
│  
│  Filename: firmware-update/NON-HLOS.bin
│  Comment: 
│  
│ +Filename: firmware-update/emmc_appsboot.mbn
│ +Comment: 
│ +
│  Filename: firmware-update/rpm.mbn
│  Comment: 
│  
│  Filename: firmware-update/sbl1.mbn
│  Comment: 
│  
│  Filename: firmware-update/sdi.mbn
│ @@ -15,14 +21,8 @@
│  
│  Filename: firmware-update/splash.img
│  Comment: 
│  
│  Filename: firmware-update/tz.mbn
│  Comment: 
│  
│ -Filename: META-INF/com/google/android/update-binary
│ -Comment: 
│ -
│ -Filename: META-INF/com/google/android/updater-script
│ -Comment: 
│ -
│  Zip file comment:
```

Roboe commented:
There's the issue with permissions and another with the ZIP's file structure. That's not new for us (https://github.com/WeAreFairphone/modem_zip_generator/issues/4#issuecomment-366345574). I don't recall how we _solved_ it, though.

z3ntu commented:
About the file order: I explicitely added the `sort` in this PR to get rid of those issues...

Roboe commented:
Yeah, that was a good idea. But it doesn't work, unfortunately.

Conclusion: ZIP is not deterministic. How we go about it?

From https://reproducible-builds.org/docs/archives/:

> ## Post-processing
>
> If tools do not support options to create reproducible archives, it is always possible to perform post-processing.
>
> [`strip-nondeterminism`](https://packages.debian.org/sid/strip-nondeterminism) already has support to normalize Zip and Jar archives (with [limitations](https://bugs.debian.org/859103)). Custom scripts like Tor Browser’s [`re-dzip.sh`](https://gitweb.torproject.org/builders/tor-browser-bundle.git/tree/gitian/build-helpers/re-dzip.sh) might also be an option.

Should we use `strip-nondeterminism`? But that would introduce yet another thing to "patch" for OSX... :confused: 

**Edit:** Oooh, I guess your machine's locale are `de-*` (or `en-*`), but mine are `es-*`. There's [a thing here](https://reproducible-builds.org/docs/archives/#file-ordering) that could improve your `sort` hack.

z3ntu commented:
Try
```diff
diff --git a/Makefile b/Makefile
index 02c8b53..7c27189 100644
--- a/Makefile
+++ b/Makefile
@@ -73,7 +73,7 @@ $(FLASHABLEZIP): $(FIRMWARE_DIR) $(EDIFY_BINARY) $(EDIFY_SCRIPT)
 	@echo "Building flashable ZIP..."
 	@mkdir -pv "$(@D)"
 	@rm -f "$@"
-	@cd "$(TEMP_DIR)" && $(FIND) -type f | $(SORT) | $(ZIP) -X \
+	@cd "$(TEMP_DIR)" && $(FIND) -type f | LC_ALL=C $(SORT) | $(ZIP) -X \
 		"$(ROOT)/$@" -@
 	@rm -rf "$(TEMP_DIR)"
 	@echo "Result: $@"
```

Roboe commented:
```
5ec0684f73637e2a8b3ba578f595b7417bb8d389a0a4197a70e25527152ba040  modem-19.02.1_2019-03-04.zip
```
Hooray!

Roboe committed:
Use locale-independent sort to fix deterministic issues with ZIP

As recommended in the Reproducible Builds's docs: https://reproducible-builds.org/docs/archives/#file-ordering

Roboe commented:
Could you try this in your MacOS machine? If it outputs the same hash as before (I guess so, but we should try it anyway), I'll consider #4 closed and we can then merge this PR and publish a new release.

z3ntu commented:
![image](https://user-images.githubusercontent.com/3768500/53751137-e42d1400-3eab-11e9-9996-c817f51bfa9d.png)

Roboe commented:
Ok, thanks! Since that's good, I'll close #4, merge this PR and publish a new release!
